samza-core/src/main/scala/org/apache/samza/storage/ContainerStorageManagerRestoreUtil.java

```diff
@@ -170,31 +170,23 @@ private static CompletableFuture<Map<TaskName, Checkpoint>> restoreAllTaskInstan
                       checkpointManager, taskRestoreManager, config, taskInstanceMetrics, executor,
                       taskBackendFactoryToStoreNames.get(taskInstanceName).get(factoryName), loggedStoreDir,
                       jobContext, containerModel);
-              try {
-                newTaskCheckpoints.put(taskInstanceName, future);
-              } catch (Exception e) {
-                String msg = String.format("DeletedException during restore task: %s after retrying to get deleted blobs.", taskName);
-                throw new SamzaException(msg, e);
-              } finally {
+              future.whenComplete((r, e) -> {
```

Comment:
This looks like an async (fire-and-forget) operation? If you do that it's possible that CSM thinks the restore managers (and stores) are closed, and recreates/reopens them, while previous close is still in progress.

Chain this to the `restoreDeletedSnapshot` future using 'handle'/ 'thenCompose' etc., and add the returned future to the map.

Reply:
Actually @dxichen suggestion is even better, since this needs to happen regardless of original or retried restore. Just chain this close operation to the future that is currently being returned/added to the map at the end of the operation, and return/add the resulting future instead.

Reply:
@prateekm that wouldn't work because returned restoreFutures only represent restores that were successfully completed,. In case of exception, new checkpoint is created and updated to a map (which is what we eventually return). So we want to make sure taskRestoreManager is closed after restoreFutures and newCheckpointFutures are completed. 
We can't return a map for restoreFutures because taskCheckpoints can be null and the concurrent map for newTaskCheckpoints can't accept that. (Also the reason why we putAll after get() on the returned map to update old map rather than replacing it)

Reply:
I will try to refractor this and simplify a bit more in a follow-up.

Reply:
Why can we not `taskRestoreFuture andThen closeTaskRestoreManager`?